File: ios/Runner/DataCloud.swift
//
//  DataCloud.swift
//  Runner
//
//  Created by Simon1 on 02/04/25.
//
import Cdp
import SFMCSDK
import MarketingCloudSDK
import Flutter

class DataCloudResult {
    var result: Bool
    var message: String
    
    init(result: Bool, message: String) {
        self.result = result
        self.message = message
    }
}


class DataCloud{
    
    var call: FlutterMethodCall
    
    init(call: FlutterMethodCall) {
        self.call = call
    }
    
    
    func initSDK() -> DataCloudResult {
        if let args = call.arguments as? [String: String],
           let appId = args["appId"],
           let endPoint = args["endPoint"] {
            
            SFMCSdk.setLogger(logLevel: LogLevel.debug)
            
            let cdpModuleConfig = CdpConfigBuilder(appId: appId, endpoint: endPoint)
                .trackScreens(true)
                .trackLifecycle(true)
                .sessionTimeout(600)
                .build()
            
            SFMCSdk.initializeSdk(ConfigBuilder().setCdp(config: cdpModuleConfig).build())
            SFMCSdk.identity.setProfileAttributes([.cdp: ["isAnonymous": "0"]])
            return DataCloudResult(result: true, message: "Success")
        } else {
            return DataCloudResult(result: false, message: "Bad Arguments")
        }
    }
    
    func login() -> DataCloudResult {
        if let args = call.arguments as? [String: String],
           let phoneNumber = args["phone_number"] {
            
            SFMCSdk.identity.setProfileAttributes([.cdp: ["phoneNumber" : phoneNumber]])
            SFMCSdk.identity.setProfileAttributes([.cdp: ["isAnonymous": "1"]])
            
            return DataCloudResult(result: true, message: "Success")
        } else {
            return DataCloudResult(result: false, message: "Bad Arguments")
        }
    }
    
    func setUserAttribute() -> DataCloudResult {
        if let args = call.arguments as? [String: String],
           let attributeName = args["attribute_name"],
           let attributeValue = args["attribute_value"] {
            let profileAttributes = [
                attributeName: attributeValue
            ]
            SFMCSdk.identity.setProfileAttributes([.cdp: profileAttributes])
            return DataCloudResult(result: true, message: "Success")
        } else {
            return DataCloudResult(result: false, message: "Bad Arguments")
        }
    }
    
    func trackEvent() -> DataCloudResult {
        guard let args = call.arguments as? [String: Any],
              let eventName = args["event_name"] as? String else {
            
            return DataCloudResult(result: false, message: "Event Name is missing")
        }
        
        let attributes = args["attributes"] as? [String: Any] ?? [:]
        
        if let customEvent = CustomEvent(name: eventName, attributes: attributes) {
            // Track the event using Salesforce SDK (or your event tracking SDK)
            SFMCSdk.track(event: customEvent)
            return DataCloudResult(result: true, message: "success")
        } else {
            return DataCloudResult(result: false, message: "failed to track event")
        }
        
    }
    
    func logout() -> DataCloudResult {
        SFMCSdk.identity.setProfileAttributes([.cdp: ["isAnonymous": "0"]])
        
        return DataCloudResult(result: true, message: "Success")
    }
    
}

File: android/app/src/main/java/com/privo/creditsaison/data_cloud/DataCloud.java
package com.privo.creditsaison.data_cloud;

import android.content.Context;
import android.util.Log;

import com.salesforce.marketingcloud.cdp.CdpSdk;
import com.salesforce.marketingcloud.cdp.consent.Consent;
import com.salesforce.marketingcloud.sfmcsdk.SFMCSdk;
import com.salesforce.marketingcloud.sfmcsdk.components.events.EventManager;
import com.salesforce.marketingcloud.sfmcsdk.modules.ModuleIdentifier;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;

public class DataCloud {

    final MethodCall call;
    final Context context;

    public DataCloud(MethodCall call, Context context) {
        this.call = call;
        this.context = context;
    }

    private static final String TAG = "DataCloud";

    public DataCloudResult login() {
        String phoneNumber = call.argument("phone_number");

        if (phoneNumber == null) {
            return new DataCloudResult(false, "Bad arguments");
        }

        try {
            SFMCSdk.requestSdk(sfmcSdk -> {
                sfmcSdk.getIdentity().setProfileAttribute("phoneNumber", phoneNumber, ModuleIdentifier.CDP);
                sfmcSdk.getIdentity().setProfileAttribute("isAnonymous", "0", ModuleIdentifier.CDP);
            });
            CdpSdk.requestSdk(cdpSdk -> cdpSdk.setConsent(Consent.OPT_IN));
            return new DataCloudResult(true, "success");
        } catch (Exception e) {
            return new DataCloudResult(false, e.getMessage());
        }

    }

    public DataCloudResult logout() {

        try {
            SFMCSdk.requestSdk(sfmcSdk -> {
                sfmcSdk.getIdentity().setProfileAttribute("isAnonymous", "1", ModuleIdentifier.CDP);
            });
            return new DataCloudResult(true, "success");
        } catch (Exception e) {
            return new DataCloudResult(false, e.getMessage());
        }

    }

    public DataCloudResult setUserAttributes() {
        String attributeName = call.argument("attributeName");
        String attributeValue = call.argument("attributeValue");

        if (attributeName == null || attributeValue == null) {
            return new DataCloudResult(false, "Bad arguments");
        }

        try {
            SFMCSdk.requestSdk(sfmcSdk -> {
                sfmcSdk.getIdentity().setProfileAttribute(attributeName, attributeValue, ModuleIdentifier.CDP);
            });
            return new DataCloudResult(true, "success");
        } catch (Exception e) {
            return new DataCloudResult(false, e.getMessage());
        }

    }

    public DataCloudResult trackEvents() {
        String eventName = call.argument("event_name");
        Map<String, ?> attributes = call.argument("attributes");

        if (eventName == null) {
            return new DataCloudResult(false, "Bad arguments");
        }

        try {

            Log.d(TAG, "trackEvents: eventName - " + eventName);
            Log.d(TAG, "trackEvents: attributes - " + attributes);

            if (attributes != null) {
                SFMCSdk.track(EventManager.customEvent(eventName, attributes));
            } else {
                SFMCSdk.track(EventManager.customEvent(eventName));
            }


            return new DataCloudResult(true, "success");
        } catch (Exception e) {
            return new DataCloudResult(false, e.getMessage());
        }

    }

}
